Add Intel TCC tutorial (#83)
canonical/real-time-ubuntu-docs: docs/tutorial/intel-tcc/preparation.md

# Preparation

This sections outlines the preparation steps for following this tutorial.
It explains which software components are required and how to set it up.

## Install Real-time Ubuntu

1. Install the latest [Ubuntu 24.04 LTS](https://releases.ubuntu.com/noble/) Desktop or Server.
   Refer to [Ubuntu Desktop installation tutorial](https://ubuntu.com/tutorials/install-ubuntu-desktop#1-overview) if needed.

2. Install the Ubuntu real-time kernel.
   Real-time Ubuntu is available via [Ubuntu Pro](https://ubuntu.com/pro/dashboard).
   The real-time kernel can be installed by running these commands:

   ```bash
   sudo pro attach
   sudo pro enable realtime-kernel
   ```

   For the real-time kernel to be used, a reboot is required.

>    We will anyway reboot when we [add kernel command-line parameters](kernel-parameters).

Reviewer: ```suggestion
   We will reboot when we [add kernel command-line parameters](kernel-parameters).
```

by the way it should be "reboot anyway" instead of "anyway reboot" :)
but upon re-reading this it's just more straightforward to say when we will reboot.
thanks.

### Install Docker and Docker Compose

```{note}
Docker may also be installed by following the [official instructions](https://docs.docker.com/engine/install/ubuntu/).
```

Install Docker and Compose as a snap by running:

```
sudo snap install docker
```

Create the docker group and add your user to it.

```
sudo addgroup --system docker
sudo adduser $USER docker
```

Restart the docker service for it to be aware of the new group.

```
sudo snap disable docker
sudo snap enable docker
```

Apply the group permissions to your current session, or log out and log back in.

```
newgrp docker
```

### Start the services

The docker-compose definition of the statistics infrastructure is provided in the example code archive you downloaded earlier.
Navigate into the `docker-compose` subdirectory, and then execute docker compose:

```bash
cd tutorial-intel-tcc-code/docker-compose
docker compose up -d 
```

Verify that the Grafana, InfluxDB, Mosquitto and Telegraf containers are up and running.
Make sure it says "Up" in the status column.

```
docker compose ps
```

If everything is running, you can connect to Grafana by following these steps:

> - Open your preferred web browser.
> - Enter the following URL in the address bar: [http://localhost:3000/](http://localhost:3000/)
>   - If you are running Docker on a remote server, replace `localhost` with the server's IP address or domain name.
>     Also make sure port 3000 is exposed through your firewall.
> 
> - Log In to Grafana: You will be prompted to log in. Use the following credentials:
>   - Username: admin
>   - Password: admin1
> 
> Browse to the statistics dashboard by going to {guilabel}`Dashboards` > {guilabel}`Provisioned Dashboards` > {guilabel}`rt_linux_tutorial`.

Reviewer: ```suggestion
1. Open your preferred web browser.
1. Enter the following URL in the address bar: [http://localhost:3000/](http://localhost:3000/)
   - If you are running Docker on a remote server, replace `localhost` with the server's IP address or domain name.
    Also make sure port 3000 is exposed through your firewall.

1. Log In to Grafana: You will be prompted to log in. Use the following credentials:
   - Username: admin
   - Password: admin1

Browse to the statistics dashboard by going to {menuselection}`Dashboards --> Provisioned Dashboards --> rt_linux_tutorial`.
The dashboard has three panels which will show "No data" at this stage.
```

Apologies @jpm-canonical ; i realized we have a better role to use `menuselection` which is more suitable here.

Also for steps that should be performed in a specific sequenced we should use numbered lists.

Author: `menuselection` sounds like the more appropriate role to use. I'll change it. The rendering is however very similar to standard text, so not very clear.

# Preparation

This sections outlines the preparation steps for following this tutorial.
It explains which software components are required and how to set it up.

> ## Install real-time Ubuntu

Reviewer: ```suggestion
## Install Real-time Ubuntu
```

any references to the 'product' would be "Real-time Ubuntu"
any references to the 'behaviour' would be "real-time".

1. Install the latest [Ubuntu 24.04 LTS](https://releases.ubuntu.com/noble/) Desktop or Server.
   Refer to [Ubuntu Desktop installation tutorial](https://ubuntu.com/tutorials/install-ubuntu-desktop#1-overview) if needed.

2. Install the Ubuntu real-time kernel.
   Real-time Ubuntu is available via [Ubuntu Pro](https://ubuntu.com/pro/dashboard).
   The real-time kernel can be installed by running these commands:

   ```bash
   sudo pro attach
   sudo pro enable realtime-kernel
   ```

   For the real-time kernel to be used, a reboot is required.

>    We will anyway reboot in a following step.

Reviewer: hi @jpm-canonical , could you please help to specify which part of the tutorial you will actually be rebooting?
it would better set the expectation and ensure that the user doesn't inadvertently miss the reboot step.
Otherwise if it doesn't do any harm i suggest just asking the user to reboot.

```suggestion
   We will be rebooting anyway when we [indicate the step].
```

   Refer to [How to enable Real-time Ubuntu](https://canonical-ubuntu-pro-client.readthedocs-hosted.com/en/latest/howtoguides/enable_realtime_kernel/) for more details.

## Other required software

Install required tools:
```bash
sudo apt install msr-tools stress-ng mosquitto-clients
```

> Download the example code archive [here](tutorial-intel-tcc-code.tar.gz).
> Then extract it.

Reviewer: ```suggestion
Download the example code archive [here](tutorial-intel-tcc-code.tar.gz), then extract it.
```

```bash
tar -xvf tutorial-intel-tcc-code.tar.gz --one-top-level
```

## Statistics infrastructure

This tutorial uses Mosquitto MQTT, Telegraf, InfluxDB and Grafana to capture and visualize the statistics obtained from the test application.

> These tools can be run as Docker containers, making the installation quicker and easier.

Reviewer: ```suggestion
We will run these tools can be run as Docker containers in this tutorial.
```

We generally avoid saying "quicker" or "easier"; that is subjective.
Also "can be run" here would indicate alternative ways, which is valid, but for the scope of the tutorial we are sticking with the docker container method.

### Install Docker and Docker Compose

```{note}
Docker may also be installed by following the [official instructions](https://docs.docker.com/engine/install/ubuntu/).
```

> Docker and Compose can be installed as a snap by running:

Reviewer: ```suggestion
Install Docker and Compose as a snap by running:
```

Similar comment to the previous. For the tutorial we are not trying to guide the user on the alternatives of how to install docker. so we can get straight to the point and say "do this".

```
sudo snap disable docker
sudo snap enable docker
```

Apply the group permissions to your current session, or log out and log back in.

```
newgrp docker
```

### Start the services

The docker-compose definition of the statistics infrastructure is provided in the example code archive you downloaded earlier.
Navigate into the `docker-compose` subdirectory, and then execute docker compose:

```bash
cd tutorial-intel-tcc-code/docker-compose
docker compose up -d 
```

Verify that the Grafana, InfluxDB, Mosquitto and Telegraf containers are up and running.
Make sure it says *Up* in the status column.

```
docker compose ps
```

If everything is running, you can connect to Grafana by following these steps:
- Open your preferred web browser.
- Enter the following URL in the address bar: [http://localhost:3000/](http://localhost:3000/)
  - If you are running Docker on a remote server, replace `localhost` with the server's IP address or domain name.
    Also make sure port 3000 is exposed through your firewall.

- Log In to Grafana: You will be prompted to log in. Use the following credentials:
  - Username: admin
  - Password: admin1

> Browse to the statistics dashboard by going to *Dashboards* > *Provisioned Dashboards* > *rt_linux_tutorial*.

Reviewer: We have the "guilabel" role. It may not be particularly helpful here but it helps with keeping the styling consistent across Ubuntu / Canonical doc sets.

```suggestion
Browse to the statistics dashboard by going to {guilabel}`Dashboards` > {guilabel}`Provisioned Dashboards` > {guilabel}`rt_linux_tutorial`.
```
